nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeduplicateRecords.java

```diff
@@ -0,0 +1,690 @@
+                    }
+
+                    recordHash = fieldValue.getValue().toString();
+                }
+
+                if (filter.contains(recordHash)) {
+                    duplicatesWriter.write(record);
+                } else {
+                    nonDuplicatesWriter.write(record);
+                    filter.put(recordHash);
+                }
+
+                index++;
+            }
+
+            // Route Non-Duplicates FlowFile
+            final WriteResult nonDuplicatesWriteResult = nonDuplicatesWriter.finishRecordSet();
+            // Route Duplicates FlowFile
+            final WriteResult duplicatesWriteResult = duplicatesWriter.finishRecordSet();
+
+            reader.close();
+            inputStream.close();
+            nonDuplicatesWriter.close();
+            nonDupeStream.close();
+            duplicatesWriter.close();
+            dupeStream.close();
+
+            final boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean();
+
+            session.adjustCounter("Records Processed",
+                    nonDuplicatesWriteResult.getRecordCount() + duplicatesWriteResult.getRecordCount(), false);
+
+            sendOrRemove(session, duplicatesFlowFile, REL_DUPLICATE, duplicatesWriter.getMimeType(),
+                    includeZeroRecordFlowFiles, duplicatesWriteResult);
+
+            sendOrRemove(session, nonDuplicatesFlowFile, REL_NON_DUPLICATE, nonDuplicatesWriter.getMimeType(),
+                    includeZeroRecordFlowFiles, nonDuplicatesWriteResult);
+
+            session.transfer(flowFile, REL_ORIGINAL);
+
+        } catch (final Exception e) {
+            logger.error("Failed in detecting duplicate records.", e);
+            session.remove(duplicatesFlowFile);
+            session.remove(nonDuplicatesFlowFile);
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private void sendOrRemove(ProcessSession session,
+                              FlowFile outputFlowFile,
+                              Relationship targetRelationship,
+                              String mimeType,
+                              boolean includeZeroRecordFlowFiles,
+                              WriteResult writeResult) {
+        if(!includeZeroRecordFlowFiles && writeResult.getRecordCount() == 0) {
+            session.remove(outputFlowFile);
+        } else {
+            Map<String, String> attributes = new HashMap<>();
+            attributes.putAll(writeResult.getAttributes());
+            attributes.put("record.count", String.valueOf(writeResult.getRecordCount()));
+            attributes.put(CoreAttributes.MIME_TYPE.key(), mimeType);
+            outputFlowFile = session.putAllAttributes(outputFlowFile, attributes);
+            if (getLogger().isDebugEnabled()) {
+                getLogger().debug("Successfully found {} unique records for {}",
+                        new Object[]{writeResult.getRecordCount(), outputFlowFile});
+            }
```

Comment:
The wrapping `isDebugEnabled()` check seems unnecessary since the arguments should not be expensive to provide to the logger method.  The wrapping `Object[]` can be removed:
```suggestion
            getLogger().debug("Processed [{}] unique records for {}", writeResult.getRecordCount(), outputFlowFile);
```

Reply:
Going to leave the isDebugEnabled in there because while the object creation is cheap, it's also unnecessary.